src/display/node_stream.js

```diff
@@ -0,0 +1,353 @@
+    this._rangeRequestReaders.push(rangeReader);
+    return rangeReader;
+  }
+
+  cancelAllRequests(reason) {
+    if (this._fullRequest) {
+      this._fullRequest.cancel(reason);
+    }
+
+    let readers = this._rangeRequestReaders.slice(0);
+    readers.forEach(function(reader) {
+      reader.cancel(reason);
+    });
+  }
+}
+
+class BaseFullReader {
+  constructor(stream) {
+    this._url = stream.url;
+    this._done = false;
+    this._errored = false;
+    this._reason = null;
+    this.onProgress = null;
+    this._length = stream.source.length;
+    this._loaded = 0;
+
+    this._fullRequest = null;
+    this._readCapability = createPromiseCapability();
+    this._headersCapability = createPromiseCapability();
+  }
+
+  get headersReady() {
+    return this._headersCapability.promise;
+  }
+
+  get contentLength() {
+    return this._length;
+  }
+
+  get isRangeSupported() {
+    return this._isRangeSupported;
+  }
+
+  get isStreamingSupported() {
+    return this._isStreamingSupported;
+  }
+
+  read() {
+    return this._readCapability.promise.then(() => {
+      if (this._done) {
+        return Promise.resolve({ value: undefined, done: true, });
+      }
+      if (this._errored) {
+        return Promise.reject(this._reason);
+      }
+
+      let chunk = this._fullRequest.read();
+      if (chunk === null) {
+        this._readCapability = createPromiseCapability();
+        return this.read();
+      }
+      this._loaded += chunk.length;
+      if (this.onProgress) {
+        this.onProgress({
+          loaded: this._loaded,
+          total: this._length,
+        });
+      }
+      return Promise.resolve({ value: chunk, done: false, });
+    });
+  }
+
+  cancel(reason) {
+    this._fullRequest.close(reason);
+    this._fullRequest.destroy(reason);
+  }
+}
+
+class BaseRangeReader {
+  constructor(stream) {
+    this._url = stream.url;
+    this._done = false;
+    this._errored = false;
+    this._reason = null;
+    this.onProgress = null;
+    this._length = stream.source.length;
+    this._loaded = 0;
+    this._readCapability = createPromiseCapability();
+
+    this._rangeRequest = null;
+  }
+
+  get isStreamingSupported() {
+    return false;
+  }
+
+  read() {
+    return this._readCapability.promise.then(() => {
+      if (this._done) {
+        return Promise.resolve({ value: undefined, done: true, });
+      }
+      if (this._errored) {
+        return Promise.reject(this._reason);
+      }
+
+      let chunk = this._rangeRequest.read();
+      if (chunk === null) {
+        this._readCapability = createPromiseCapability();
+        return this.read();
+      }
+      this._loaded += chunk.length;
+      if (this.onProgress) {
+        this.onProgress({
+          loaded: this._loaded,
+          total: this._length,
+        });
+      }
+      return Promise.resolve({ value: chunk, done: false, });
+    });
+  }
+
+  cancel(reason) {
+    this._rangeRequest.cancel(reason);
```

Comment:
`_rangeRequest` should not be a property of the base class, because the base class does not need to know about the implementation detail, so you should remove the use of `_rangeRequest` here. Looking at the implementation, you should probably not directly expose `_rangeRequest`, but possibly store a reference to the `.read` method, maybe as `_read`.  
Besides, the `_rangeRequest` object is not the same in the subclasses:

The argument of the callback to `http.request` is an IncomingMessage, which does not have a `cancel` method - https://nodejs.org/api/http.html#http_class_http_incomingmessage

In PDFNodeStreamFsRangeReader, the type is a ReadableStream, which does not have a `cancel` method either.